### PR TITLE
Feat/content negotiation

### PR DESCRIPTION
**Pull request by Acrontum-Carmichael**
Relates to issues:
https://github.com/acrontum/generate-it-typescript-server/issues/36
&
https://github.com/acrontum/generate-it-typescript-server/issues/41

Essentially the content output type is handled by a middleware. The use cases are slim but this flexibility now allows the developer to (for example) deliver cached js or html from the API. Previously it was all only json / sendFile or download file.

Based on @p-mcgowan previous work to pull out the most rel content type from an input type and given output options, this is now used in the above said middleware to act on the calculated content-type.

It is a fairly simple concept.. the api file defines the available output format (eg ['text/html', 'text/plain']). The client makes a request to a route with the "Accept" header (https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept). If the requested type is available to output happy days.. if not.. http 406 is thrown..

If no accept header is provided or the accept is simply `*/*` then the 1st permitted type in the defined list from the openapi file is used.

The nice thing is now the routes/__op file is much simpler and the middleware now allows for much easier advancement for more complex output calculations.

### CODE REVIEW

**Comment by Acrontum-Carmichael** [comment]
Docs required still.

this get:
```
tags:
  - {{ autoTag() }}
summary: Fetches the file for a given URL ID
operationId: {{ uniqueOpId() }}
produces:
  - application/pdf
responses:
  '200':
    description: Success
```

Works perfectly fine.. the only thing to remember is that all download type required the domain return a file path string

File: src/http/nodegen/middleware/inferResponseType.ts [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+      if (dataOrPath === undefined) {
+        return res.sendStatus(status);
+      }
+
+      permittedTypes = Array.isArray(permittedTypes) ? permittedTypes : [permittedTypes];
+
+      // Calculate the calculatedAcceptHeader based on the provided accept header
+      // When the accept header is not provided then we fallback to the 1st permittedType
+      // The last fallback is then application/json
+      const calculatedAcceptHeader = req.headers['accept'] ?
```
Comment (p-mcgowan) [inline src/http/nodegen/middleware/inferResponseType.ts:27]: `calculatedContentType` would be more accurate - and makes more below where `return res.set('Content-Type', calculatedAcceptHeader)`

**Review by p-mcgowan — APPROVED** [review]
Looks good to me - left a comment on a variable name but that's not a big deal.

I like how much that cleans up the template.

What was the driving force behind using `res.inferResponseType` instead of calling an imported method?

**Comment by Acrontum-Carmichael** [comment]
The driving force behind adding this as a middleware was so there are fewer moving parts in the custom code, also that the req and res object are automatically available in a middleware so a little less code is generated.

As we discussed @p-mcgowan the extra addition here is that the http layer is now 100% managed with the option for injection from the app.ts for those who want to inject their own middlewares.

@acrontum-redjol this is a breaking change to existing code but the server.ts and app.ts just need to be manually updated but the change is trivial compared to the gain this brings.

**Review by p-mcgowan — APPROVED** [review]
Yep lets give it a whirl and see how it goes - there are quite a fee changes to go over and it won't be clear what effect that has until we throw it into some projects

**Pull request merged**